ARROW-16253: [R] Helper function for casting from float to duration via int64()

dragosmg opened the pull request.

jonkeane reviewed (commented):
Looking through, I think I found a few other places this might be helpful. What about the followin?

https://github.com/apache/arrow/blob/3a646b35b29e52ca8eca995eb83af29c44b191c4/r/R/dplyr-funcs-datetime.R#L415-L417

https://github.com/apache/arrow/blob/3a646b35b29e52ca8eca995eb83af29c44b191c4/r/R/dplyr-funcs-datetime.R#L354-L355

We also mention that issue at https://github.com/apache/arrow/blob/3a646b35b29e52ca8eca995eb83af29c44b191c4/r/R/dplyr-datetime-helpers.R#L146 is it possible to extend this helper to help there too? (plus any other of the places we see this same pattern?)

dragosmg commented:
Thanks @Jon. I should've probably offered some details on why I didn't use `make_duration()` in those places.

https://github.com/apache/arrow/blob/3a646b35b29e52ca8eca995eb83af29c44b191c4/r/R/dplyr-funcs-datetime.R#L415-L417

I had added a comment in https://github.com/apache/arrow/pull/13055/commits/b4ccc08125e015b3fdb901ab3f796e3a3ca379fc, on why `make_duration()` cannot be used there. We need the casting chain for a second reason, i.e. to get the units right. Switching to `make_duration()` would involve adding another transformation to account for the change in measurement unit. There is a relevant conversation over at [ARROW-15996](https://issues.apache.org/jira/browse/ARROW-15996).

for 

https://github.com/apache/arrow/blob/3a646b35b29e52ca8eca995eb83af29c44b191c4/r/R/dplyr-funcs-datetime.R#L354-L355

Done! That was an oversight

for 

https://github.com/apache/arrow/blob/3a646b35b29e52ca8eca995eb83af29c44b191c4/r/R/dplyr-datetime-helpers.R#L146

we can't really use `make_duration()` there (as only the cast to `int64()`, and not the cast to `duration("s")` is needed) so I deleted the comment referencing [ARROW-16253](https://issues.apache.org/jira/browse/ARROW-16253).

dragosmg commented:
I think this is ready for another look.

jonkeane commented on r/R/dplyr-funcs-datetime.R:
```diff
@@ -412,8 +412,9 @@ register_bindings_duration <- function() {
+      # complex casting due to cast type restrictions: time64 -> int64 -> duration(us)
       # and then we cast to duration ("s") at the end
+      # we also need the casting chain to get the measurement units right
       x <- x$cast(time64("us"))$cast(int64())$cast(duration("us"))
```
could this be the following?

```suggestion
      x <- make_duration(x$cast(time64("us"), unit = "us")
```

jonkeane commented on r/R/dplyr-funcs-datetime.R:
```diff
@@ -412,9 +412,10 @@ register_bindings_duration <- function() {
       x <- build_expr("strptime", x, options = list(format = format, unit = 0L))
-      # complex casting only due to cast type restrictions: time64 -> int64 -> duration(us)
+      # complex casting due to cast type restrictions: time64 -> int64 -> duration(us)
       # and then we cast to duration ("s") at the end
-      x <- x$cast(time64("us"))$cast(int64())$cast(duration("us"))
+      # we also need the casting chain to get the measurement units right
+      x <- make_duration(x$cast(time64("us")), unit = "us")
```
Would you mind updating this comment to match what we have now? Since we've encapsulated the `make_duration()` complexities we don't have to worry about those. Though we might still want to keep that we're  going to `s` later on.

jonkeane reviewed (approved):
This is looking good to go (with a slight clean up of a comment).

I've not looked extensively to see if there are other places we could use this like the handful I noted, have you taken a look through the date-time codebase to see if there are more?

dragosmg commented:
I haven't found more places in which to use this.
